feat(github): Autofill repository name when setting up GitHub builds

gajeshbhat opened the pull request:
## Done

Implemented autofill functionality for the RepoSelector component. When a user selects an organization, the repository input automatically fills with a matching repository name if one exists that matches the snap name.

**Key features:**
- Automatic repository detection when an organization is selected
- Smart autofill when `repo.name === snapId`
- Automatic validation after autofill
- User can override autofilled values
- Performance optimized with proper memoization
- Graceful error handling and loading states

## How to QA

### Setup Requirements

**Environment Configuration:**
Create `.env.local` with GitHub OAuth and Launchpad credentials:

```bash
# GitHub OAuth
GITHUB_CLIENT_ID=your_client_id
GITHUB_CLIENT_SECRET=your_client_secret

# Launchpad
LP_API_URL=https://api.launchpad.net/devel
LP_API_USERNAME=your_username
LP_API_CONSUMER_KEY=build.snapcraft.io.development
LP_API_TOKEN=your_token
LP_API_TOKEN_SECRET=your_secret
```

**Test Data:**
- Register a snap on snapcraft.io
- Create a GitHub repository with the same name as your snap
- Ensure you have repositories with/without `snapcraft.yaml` for testing

### API Testing

**Test GitHub Integration:**
```bash
dotrun
Open in browser "http://localhost:8004/publisher/github/get-repos"
# Expected: JSON array of repositories
```

**Test Repository Validation:**
```bash
Open in browser "http://localhost:8004/api/snap-name/builds/validate-repo?repo=owner/repo"
# Expected: JSON with validation result
```

### Manual Testing

**Test Scenarios:**

1. **Successful Autofill:**
   - Navigate to `/snap-name/builds` where you have a matching repository
   - Select organization → input should autofill → validation runs → "Start building" appears

2. **Autofill with Error:**
   - Same as above but with repository missing `snapcraft.yaml`
   - Should show error message and "Re-check" button

3. **No Autofill:**
   - Use snap name with no matching repository
   - Input should remain empty, manual selection should work

4. **User Override:**
   - Let autofill happen → clear input → type different repo name
   - Manual selection should work normally

5. **Organization Switching:**
   - Select org A → autofill → switch to org B → input clears → new autofill attempt

**Expected Results:**
- Autofill only when the repository name matches snap name
- Automatic validation after autofill
- Clear error messages for validation failures
- User can override autofilled values

## Testing
- [x] This PR has tests
- [ ] No testing required (explain why):

**Test Coverage:**
- 11 comprehensive unit tests covering all autofill scenarios
- Tests for autofill success/failure, manual override, organization switching, validation states
- All tests pass (11/11), no linting or TypeScript errors

## Issue / Card
Fixes #3319
Internally WD-26505

## Screenshots

Two screencasts have been attached showing:
1. Organization selection triggering autofill
2. Automatic validation running
3. Success state with "Start building" button
4. Error state with validation message
5. Manual override functionality

[Snapcraft-io-autofill-repo-not-present-3319.webm](https://github.com/user-attachments/assets/365a6d9d-1df9-412a-9864-735c81bd3d80)

[snapcraft-io-github-repo-autofill-3319.webm](https://github.com/user-attachments/assets/8da04c3a-0a54-459c-a256-8915b3445d51)

gajeshbhat commented:
> ## [Codecov](https://app.codecov.io/gh/canonical/snapcraft.io/pull/5347?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=canonical) Report
> ❌ Patch coverage is `13.33333%` with `13 lines` in your changes missing coverage. Please review. ✅ Project coverage is 59.07%. Comparing base ([`b8b6b55`](https://app.codecov.io/gh/canonical/snapcraft.io/commit/b8b6b5508319cdaea66d4c80420a503cd3a20c24?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=canonical)) to head ([`4ff39cd`](https://app.codecov.io/gh/canonical/snapcraft.io/commit/4ff39cd3eea727e6bd92ef0eb5be718d85b568df?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=canonical)). ⚠️ Report is 485 commits behind head on main.
> 
> [Files with missing lines](https://app.codecov.io/gh/canonical/snapcraft.io/pull/5347?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=canonical)	Patch %	Lines
> [static/js/publisher/pages/Builds/RepoSelector.tsx](https://app.codecov.io/gh/canonical/snapcraft.io/pull/5347?src=pr&el=tree&filepath=static%2Fjs%2Fpublisher%2Fpages%2FBuilds%2FRepoSelector.tsx&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=canonical#diff-c3RhdGljL2pzL3B1Ymxpc2hlci9wYWdlcy9CdWlsZHMvUmVwb1NlbGVjdG9yLnRzeA==)	13.33%	[12 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/canonical/snapcraft.io/pull/5347?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=canonical)
> Additional details and impacted files
> 🚀 New features to boost your workflow:

Will add some Unit tests to improve coverage.

gajeshbhat commented:
Finished adding unit tests for the autofill feature.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -143,9 +166,13 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+
     if (!searchTerm) {
       setValidationError(false);
       setValidRepo(false);
```
Fixed - changed setValidRepo(false) to setValidRepo(null) when input is empty to correctly represent "not validated" state.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -35,44 +42,72 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+        setValidating(false);
+      }
+    },
+    [snapId, getRepoNameWithOwner],
```
Fixed. Memoized `getRepoNameWithOwner` function

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -117,47 +172,64 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
-    }
+      if (!searchTerm) {
+        setValidationError(false);
+        setValidRepo(false);
```
Updated - testing to see if Co-Pilot reverts to the original warning of using null

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -35,44 +43,90 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
-      setValidationError(true);
-      setValidRepo(null);
+  // Autofill effect with proper dependencies
+  useEffect(() => {
+    const currentOrgKey = `${selectedOrg || "user"}-${snapId}`;
+
+    if (
+      repos.length > 0 &&
+      snapId &&
+      !repoInputValue &&
+      autofillAttemptedRef.current !== currentOrgKey
+    ) {
+      const matchingRepo = repos.find((repo: Repo) => repo.name === snapId);
+      if (matchingRepo) {
+        setRepoInputValue(matchingRepo.name);
+        setSelectedRepo(matchingRepo);
+        validateRepoInternal(matchingRepo);
+      }
+      // Mark autofill as attempted for this org/snap combination
+      autofillAttemptedRef.current = currentOrgKey;
     }
+  }, [repos, snapId, repoInputValue, selectedOrg, validateRepoInternal]);
```
The autofill validation is intentional and occurs only once per organization selection, not during rapid user input. Since users expect immediate feedback when autofill occurs, and this represents a single API call per deliberate user action (not continuous typing), debouncing would actually degrade the user experience without providing meaningful performance benefits.

gajeshbhat commented:
@bartaz I went back and forth with Copilot for a while and made some improvements based on its suggestion. Please let me know if you have any questions. Cheers.

bartaz commented:
> @bartaz I went back and forth with Copilot for a while and made some improvements based on its suggestion. Please let me know if you have any questions. Cheers.

Thanks @gajeshbhat for your contribution. As we have noticed your increasing activity across our repos we want to ensure we give the pull requests a proper and thorough review. Given our current priorities, we can't guarantee immediate reviews. Thanks for your patience and understanding. We will get back to this when we can.

In the meantime, it would be helpful if you would follow our [PR description template](https://github.com/canonical/snapcraft.io/blob/main/.github/pull_request_template.md), including the QA steps and screenshots/screencasts where relevant to help us get context on the proposed changes and properly test them.

gajeshbhat commented:
> > @bartaz I went back and forth with Copilot for a while and made some improvements based on its suggestion. Please let me know if you have any questions. Cheers.
> 
> 
> 
> Thanks @gajeshbhat for your contribution. As we have noticed your increasing activity across our repos we want to ensure we give the pull requests a proper and thorough review. Given our current priorities, we can't guarantee immediate reviews. Thanks for your patience and understanding. We will get back to this when we can.
> 
> 
> 
> In the meantime, it would be helpful if you would follow our [PR description template](https://github.com/canonical/snapcraft.io/blob/main/.github/pull_request_template.md), including the QA steps and screenshots/screencasts where relevant to help us get context on the proposed changes and properly test them.

Thank you @bartaz for taking the time to look at my PR. I totally understand that you might have higher priority items to work on and this definitely is a low priority feature. I will follow the contributors guidelines more closely and post screenshots and screen recording of testing this feature so that's it's easy to review. I think that's a very reasonable ask. I have previously done something similar for PR FOR Ubuntu App Centre and would be happy to repeat the same process here.

As for my increased activity, I have been using canonical software for well over a decade now but for the longest time haven't had the time, resources and skills to contribute. With time, this has changed and this summer I decided to dedicate some time to give back to community. Some of my PRs might not be the highest quality as I am still new to these code bases, I hope it improves over time.

Thanks again. Cheers.

gajeshbhat commented:
Rebased my PR.

gajeshbhat commented:
https://github.com/canonical/snapcraft.io/actions/runs/17478582923/job/49644201477?pr=5347 -> test-js CI chek fialing after rebasing with main. Was passing earlier. Not related to my change.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -81,20 +135,24 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+      const responseData = await response.json();
+      setRepos(responseData);
+    } catch (error) {
+      console.error("Failed to fetch repos:", error);
```
Fixed! Replaced console.error with user-friendly error message using p-form-validation__message pattern as suggested.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -166,17 +241,28 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+        setRepoConnected(true);
+      }
+    } catch (error) {
+      console.error("Failed to connect repo:", error);
```
Fixed! Replaced console.error with proper form validation error display following the canonical react-components pattern.

gajeshbhat commented:
> This works quite nicely. Just a couple of comments in the code.
> 
> There is an issue which wasn't introduced by this PR, but the "Re-check" button when the wrong snap is selected has no icon. For this to work the icon would need to be [imported here](https://github.com/canonical/snapcraft.io/blob/292295688de13ff17d0917a2aeb795a0545b60ed/static/sass/styles.scss#L179) e.g. `@include vf-p-icon-restart;` and if the text in the tooltip could be changed to "Check again" rather than "Re-check".
> 
> It would be appreciated if you could make that change also 🙏

Thank you for the review @steverydz . I have updated the code to fix the **Check again** button as you asked. Below are the screenshots and the screencast that demonstrate these changes.

<img width="2477" height="941" alt="Recheck Button Fix" src="https://github.com/user-attachments/assets/6416a94f-6308-47bb-b7c3-0abb6695af84" />

<img width="2493" height="1009" alt="Recheck Button Tooltip" src="https://github.com/user-attachments/assets/d44633d8-351f-467d-b2be-6f3d53e27231" />

<img width="2490" height="933" alt="Recheck Fix Working Green" src="https://github.com/user-attachments/assets/383e50ed-df3c-47dc-b0c7-af91dca1b38b" />

[Snapcraft-io-recheck-button.webm](https://github.com/user-attachments/assets/eec040a0-bcc3-48fe-9d6a-4b9ca50278b9)

gajeshbhat commented:
Fixed tests and lint errors; all CI checks now pass.

bartaz reviewed (commented):
Some questions around `useCallback` usage.

bartaz commented on static/js/publisher/pages/Builds/__tests__/RepoSelector.test.tsx:
```diff
@@ -0,0 +1,406 @@
+      const repoInput = screen.getByLabelText(
+        "Select a repository",
+      ) as HTMLInputElement;
+      expect(repoInput.value).toBe("test-snap-id");
```
How do we know it was cleared and filled again? If the value is exactly the same as in previous `expect` it could have not changed at all.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -35,44 +43,92 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
-    const response = await fetch(
-      `/api/${snapId}/builds/validate-repo?repo=${repoName}`,
-    );
+  const getRepoNameWithOwner = useCallback(
```
I've extracted it as a proper utility function outside the component that takes both org and repo as parameters, eliminating both the unnecessary memoization and avoiding any dependency on component state. Thanks for the feedback.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -35,44 +43,92 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+      setValidating(true);
+
+      // Calculate repo name inline to avoid dependency on getRepoNameWithOwner
+      const repoName = selectedOrg
```
I've now refactored this and updated the code.

gajeshbhat commented:
> Thanks for the updates @gajeshbhat. If I could have one request for your future contributions, it would be helpful if you just push individual commits when you update things in PR, especially once its in review. Otherwise it's a bit hard to track what changes you are introducing. We will still squash it before merging, so no need to force push single comment on the PR. Thanks!
> 
> I'm in the process of testing and reviewing this myself, and should get back to you later today or tomorrow.

Thanks for taking a look @bartaz . No worries. Will push individual commits from now on.

gajeshbhat commented:
> Some questions around `useCallback` usage.

Thanks for pointing it out and being patient with me as I work to improve this. I have addressed your comments and made some changes. Please let me know if you have any further questions. Appreciate your feedback to help improve the code.

bartaz reviewed (commented):
Thanks for your responses and updates. There are still few things I commented on for clarifications or potential improvements, but we are getting there.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -35,43 +47,76 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+  };
 
-    setValidating(false);
+  const validateRepo = async (repo: Repo | undefined) => {
```
I originally added it to keep the null check separate from the validation logic. I made some changes and pushed the commit. Let me know what you think.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -35,43 +47,76 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
     setValidating(true);
 
-    const repoName = getRepoNameWithOwner(repo);
+    const repoName = selectedOrg
```
I  kept the conditional logic outside because I wanted `getRepoNameWithOwner` to be a pure utility function that just formats strings. I refactored this to handle the `selectedOrg` check internally. Let me know what you think about it.

gajeshbhat commented:
Also made a simple one-line change to remove redundant `setReposLoading(true)` when changing organizations, as this is also called under `getRepos()`. The performance gain from this duplication is likely not worth it (1-2 milliseconds).

steverydz reviewed (approved):
LGTM 👍

gajeshbhat commented:
> LGTM 👍

Thank you, @steverydz, for accepting my contribution.

gajeshbhat commented on static/js/publisher/pages/Builds/RepoSelector.tsx:
```diff
@@ -67,55 +82,79 @@ function RepoSelector({ githubData, setAutoTriggerBuild }: Props) {
+  // Track autofill attempts to avoid unnecessary re-runs
+  const autofillAttemptedRef = useRef<string | null>(null);
 
+  const validateRepoInternal = async (repo: Repo | undefined) => {
```
I can put up a PR making this tiny change, as I understand how this can cause confusion. Should be a short PR. What do you think?

gajeshbhat commented:
> Thanks @gajeshbhat, this is now merged.

Thank you @bartaz for the review and guidance.
